sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/ShuffleExchangeExec.scala

```diff
@@ -47,6 +47,15 @@ import org.apache.spark.util.random.XORShiftRandom
  */
 trait ShuffleExchangeLike extends Exchange {
 
+  /**
+   * The asynchronous job that materializes the shuffle. It also does the preparations work,
+   * such as waiting for the subqueries.
+   */
+  @transient private lazy val shuffleFuture: Future[MapOutputStatistics] = executeQuery {
+    materializationStarted.set(true)
```

Comment:
After a closer look, I don't think this change works as we expect. We set this `materializationStarted` flag before we return the `Future`, which means we are still on the AQE loop's main thread. That said, once we submit a query stage, its `materializationStarted` becomes true immediately and we can't really avoid the wasted query stage execution.

The test passed because `ShuffleExchangeExec` calls `child.execute()` before returning the `Future`. Then we exit the AQE loop without cancelling other stages.

Reply:
I think we need a bit of synchronization here. The shuffle node should have two fields: `isCancelled` flag and the shuffle job `Future`.
- When we cancel a shuffle, we lock on the shuffle node, and set `isCancelled` flag to true. Then if the shuffle job `Future` is present, we cancel it.
- When we are going to submit a shuffle, we lock on the shuffle node. Then: if the `isCancelled` flag is true, fail immediately, otherwise, submit the shuffle job and set the `Future` field.

Reply:
cc @erenavsarogullari  @ulysses-you

Reply:
It seems not an acutally issue for now. AQE always do materilize stage and cancel stage at main thread. So if we decide to cancel stage then that means we will never do materilize stage again. We may need improve this code if we support do materilize concurrently in future.